File: Ocaml/abb_program.ml
open Abb

let rec preorder (t: tBST) = match (!^t) with
    Empty ->
        Printf.printf " () "
|   Node {key = k; left = l; right = r} ->
        if (isEmptyTree l) && (isEmptyTree r) then
                (Printf.printf " ( ";
                Printf.printf "%d" k;
                Printf.printf " ) ")
        else
            (Printf.printf " ( ";
            Printf.printf "%d " k;
            preorder l;
            preorder r;
            Printf.printf " ) ");;

let t = new_pointer Empty;;
insertKey t 4;;
insertKey t 4;;
insertKey t 2;;
insertKey t 6;;
insertKey t 1;;
insertKey t 3;;
insertKey t 5;;
insertKey t 7;;

preorder t;; Printf.printf "\n";;

Printf.printf "Searching for 1... %d \n" (root (searchKey t 1));;
Printf.printf "Searching for 2... %d \n" (root (searchKey t 2));;
Printf.printf "Searching for 3... %d \n" (root (searchKey t 3));;
Printf.printf "Searching for 4... %d \n" (root (searchKey t 4));;
Printf.printf "Searching for 5... %d \n" (root (searchKey t 5));;
Printf.printf "Searching for 6... %d \n" (root (searchKey t 6));;
Printf.printf "Searching for 7... %d \n" (root (searchKey t 7));;

Printf.printf "Deleting 5...\n";;
removeKey t 5;;
preorder t;; Printf.printf "\n";;

Printf.printf "Deleting 6...\n";;
removeKey t 6;;
preorder t;; Printf.printf "\n";;

Printf.printf "Deleting 4...\n";;
removeKey t 4;;
preorder t;; Printf.printf "\n";;

Printf.printf "Deleting 2...\n";;
removeKey t 2;;
preorder t;; Printf.printf "\n";;

File: Python/abb_program.py
from abb import *

def preorder(bst):
	preorder_node(bst.root)
	print("")

def preorder_node(bst):
	print(" ( ", end="")
	if not isEmpty(bst.get()):
		if not isEmpty(bst.get().leftChild.get()) or not isEmpty(bst.get().rightChild.get()):
			print(bst.get().key, end="")
			preorder_node(bst.get().leftChild)
			preorder_node(bst.get().rightChild)
		else:
			print(bst.get().key, end="")
	print(" ) ", end="")

T = BinarySearchTree()
insertKey(T, 4)
insertKey(T, 4)
insertKey(T, 2)
insertKey(T, 6)
insertKey(T, 1)
insertKey(T, 3)
insertKey(T, 5)
insertKey(T, 7)

preorder(T)

print("Search 1...", str(root(searchKey(T, 1))))
print("Search 2...", str(root(searchKey(T, 2))))
print("Search 3...", str(root(searchKey(T, 3))))
print("Search 4...", str(root(searchKey(T, 4))))
print("Search 5...", str(root(searchKey(T, 5))))
print("Search 6...", str(root(searchKey(T, 6))))
print("Search 7...", str(root(searchKey(T, 7))))

print("Remove 5...", end="")
removeKey(T, 5)
preorder(T)
print("Remove 6...", end="")
removeKey(T, 6)
preorder(T)
print("Remove 4...", end="")
removeKey(T, 4)
preorder(T)
print("Remove 2...", end="")
removeKey(T, 2)
preorder(T)